## Part 1: Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [2]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [3]:
# Check to see if dataframe contain any nulls
display(attrition_df.isnull().sum())

Age                         0
Attrition                   0
BusinessTravel              0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EnvironmentSatisfaction     0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
NumCompaniesWorked          0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSinceLastPromotion     0
YearsWithCurrManager        0
dtype: int64

In [4]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[['Attrition', 'Department']]
y_df.head()


Attrition              Department
0       Yes                   Sales
1        No  Research & Development
2       Yes  Research & Development
3        No  Research & Development
4        No  Research & Development

In [5]:
# List column names in dataframe
attrition_df.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EnvironmentSatisfaction', 'HourlyRate',
       'JobInvolvement', 'JobLevel', 'JobRole', 'JobSatisfaction',
       'MaritalStatus', 'NumCompaniesWorked', 'OverTime', 'PercentSalaryHike',
       'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [6]:
# Create a list of at least 10 column names to use as X data
selected_cols = ['Age', 'Education', 'EducationField',
                 'JobRole', 'JobSatisfaction', 'PerformanceRating',
                 'TotalWorkingYears', 'TrainingTimesLastYear',
                 'YearsAtCompany', 'YearsInCurrentRole',
                 'YearsSinceLastPromotion']

# Create X_df using your selected columns
X_df = attrition_df[selected_cols]
display(X_df.head())

# Show the data types for X_df
display(X_df.dtypes)



Age  Education EducationField                JobRole  JobSatisfaction  \
0   41          2  Life Sciences        Sales Executive                4   
1   49          1  Life Sciences     Research Scientist                2   
2   37          2          Other  Laboratory Technician                3   
3   33          4  Life Sciences     Research Scientist                3   
4   27          1        Medical  Laboratory Technician                2   

   PerformanceRating  TotalWorkingYears  TrainingTimesLastYear  \
0                  3                  8                      0   
1                  4                 10                      3   
2                  3                  7                      3   
3                  3                  8                      3   
4                  3                  6                      3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  
0               6                   4                        0  
1              10                   7                        1  
2               0                   0                        0  
3               8                   7                        3  
4               2                   2                        2

Age                         int64
Education                   int64
EducationField             object
JobRole                    object
JobSatisfaction             int64
PerformanceRating           int64
TotalWorkingYears           int64
TrainingTimesLastYear       int64
YearsAtCompany              int64
YearsInCurrentRole          int64
YearsSinceLastPromotion     int64
dtype: object

In [7]:
# Display X_df column values
for cols in X_df.columns:
  print(f"-------------- {cols} --------------")
  print(X_df[cols].unique())
  print("--------------------------------------------")
  print("\n")

-------------- Age --------------
[41 49 37 33 27 32 59 30 38 36 35 29 31 34 28 22 53 24 21 42 44 46 39 43
 50 26 48 55 45 56 23 51 40 54 58 20 25 19 57 52 47 18 60]
--------------------------------------------


-------------- Education --------------
[2 1 4 3 5]
--------------------------------------------


-------------- EducationField --------------
['Life Sciences' 'Other' 'Medical' 'Marketing' 'Technical Degree'
 'Human Resources']
--------------------------------------------


-------------- JobRole --------------
['Sales Executive' 'Research Scientist' 'Laboratory Technician'
 'Manufacturing Director' 'Healthcare Representative' 'Manager'
 'Sales Representative' 'Research Director' 'Human Resources']
--------------------------------------------


-------------- JobSatisfaction --------------
[4 2 3 1]
--------------------------------------------


-------------- PerformanceRating --------------
[3 4]
--------------------------------------------


-------------- TotalWorkingYea

In [8]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_df, y_df, random_state=42)

# Display X_train, X_test, y_train, y_test after splitting
display(X_train.head())
display(X_train.shape)

display(y_train.head())
display(y_train.shape)

display(X_test.head())
display(X_test.shape)

display(y_test.head())
display(y_test.shape)


Age  Education EducationField                JobRole  JobSatisfaction  \
1343   29          3  Life Sciences  Laboratory Technician                1   
1121   36          4  Life Sciences        Sales Executive                3   
1048   34          3          Other        Sales Executive                1   
1393   27          3      Marketing        Sales Executive                4   
527    32          3      Marketing        Sales Executive                4   

      PerformanceRating  TotalWorkingYears  TrainingTimesLastYear  \
1343                  3                 11                      2   
1121                  3                 15                      5   
1048                  3                 15                      2   
1393                  3                  7                      5   
527                   3                 10                      2   

      YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  
1343               3                   2                        1  
1121               1                   0                        0  
1048              13                   9                        3  
1393               7                   7                        0  
527               10                   7                        0

(1102, 11)

Attrition              Department
1343        No  Research & Development
1121        No                   Sales
1048        No                   Sales
1393        No                   Sales
527         No                   Sales

(1102, 2)

Age  Education   EducationField                 JobRole  \
1041   28          3          Medical         Sales Executive   
184    53          2          Medical  Manufacturing Director   
1222   24          1  Human Resources         Human Resources   
67     45          3    Life Sciences      Research Scientist   
220    36          2    Life Sciences   Laboratory Technician   

      JobSatisfaction  PerformanceRating  TotalWorkingYears  \
1041                1                  3                  6   
184                 1                  3                  5   
1222                3                  3                  1   
67                  1                  3                 25   
220                 2                  3                 16   

      TrainingTimesLastYear  YearsAtCompany  YearsInCurrentRole  \
1041                      4               5                   4   
184                       3               4                   2   
1222                      2               1                   0   
67                        2               1                   0   
220                       3              13                  11   

      YearsSinceLastPromotion  
1041                        1  
184                         1  
1222                        0  
67                          0  
220                         3

(368, 11)

Attrition              Department
1041        No                   Sales
184         No  Research & Development
1222       Yes         Human Resources
67          No  Research & Development
220         No  Research & Development

(368, 2)

In [9]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary

from sklearn.preprocessing import LabelEncoder, OneHotEncoder

# Preprocess X_train "EducationField" and "JobRole" columns (one-hot encoding)
categorical_cols = ['EducationField', 'JobRole']

# Create Encoder
categorical_encoder = OneHotEncoder(drop='first', sparse_output=False)

# Fit and transform X_train categorical columns
categorical_encoded = categorical_encoder.fit_transform(X_train[categorical_cols])
#display(categorical_encoded)

# Get Encoded Column names
enc_categorical_cols = categorical_encoder.get_feature_names_out(categorical_cols)
#display(enc_categorical_cols)

# Create Encoded dataframe
df_categorical_encoded = pd.DataFrame(categorical_encoded, columns=enc_categorical_cols)
#display(df_categorical_encoded.head())

# Reindex X_train before concatenate to align with df_categorical_encoded DataFrame
X_train_reindexed = X_train.reset_index(drop=True)

# Concatenate the encoded columns to the original DataFrame
X_train_encoded = pd.concat([X_train_reindexed, df_categorical_encoded], axis=1)

# Drop the original "EducationField" and "JobRole" columns
X_train_encoded = X_train_encoded.drop(['EducationField', 'JobRole'], axis=1)

# Display Dataframe and shape
display(X_train_encoded.head())
display(X_train_encoded.shape)

# Check to see if dataframes contain any nulls
#display(X_train_encoded.isnull().sum())


Age  Education  JobSatisfaction  PerformanceRating  TotalWorkingYears  \
0   29          3                1                  3                 11   
1   36          4                3                  3                 15   
2   34          3                1                  3                 15   
3   27          3                4                  3                  7   
4   32          3                4                  3                 10   

   TrainingTimesLastYear  YearsAtCompany  YearsInCurrentRole  \
0                      2               3                   2   
1                      5               1                   0   
2                      2              13                   9   
3                      5               7                   7   
4                      2              10                   7   

   YearsSinceLastPromotion  EducationField_Life Sciences  ...  \
0                        1                           1.0  ...   
1                        0                           1.0  ...   
2                        3                           0.0  ...   
3                        0                           0.0  ...   
4                        0                           0.0  ...   

   EducationField_Other  EducationField_Technical Degree  \
0                   0.0                              0.0   
1                   0.0                              0.0   
2                   1.0                              0.0   
3                   0.0                              0.0   
4                   0.0                              0.0   

   JobRole_Human Resources  JobRole_Laboratory Technician  JobRole_Manager  \
0                      0.0                            1.0              0.0   
1                      0.0                            0.0              0.0   
2                      0.0                            0.0              0.0   
3                      0.0                            0.0              0.0   
4                      0.0                            0.0              0.0   

   JobRole_Manufacturing Director  JobRole_Research Director  \
0                             0.0                        0.0   
1                             0.0                        0.0   
2                             0.0                        0.0   
3                             0.0                        0.0   
4                             0.0                        0.0   

   JobRole_Research Scientist  JobRole_Sales Executive  \
0                         0.0                      0.0   
1                         0.0                      1.0   
2                         0.0                      1.0   
3                         0.0                      1.0   
4                         0.0                      1.0   

   JobRole_Sales Representative  
0                           0.0  
1                           0.0  
2                           0.0  
3                           0.0  
4                           0.0  

[5 rows x 22 columns]

(1102, 22)

In [10]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary

# Preprocess X_test "EducationField" and "JobRole" columns (one-hot encoding)
# Transform X_test categorical columns
test_categorical_encoded = categorical_encoder.transform(X_test[categorical_cols])
#display(test_categorical_encoded)

# Create Encoded dataframe
df_test_categorical_encoded = pd.DataFrame(test_categorical_encoded, columns=enc_categorical_cols)
#display(df_test_categorical_encoded.head())

# Reindex X_test before concatenate to align with df_categorical_encoded DataFrame
X_test_reindexed = X_test.reset_index(drop=True)

# Concatenate the encoded columns to the original DataFrame
X_test_encoded = pd.concat([X_test_reindexed, df_test_categorical_encoded], axis=1)

# Drop the original "EducationField" and "JobRole" columns
X_test_encoded = X_test_encoded.drop(['EducationField', 'JobRole'], axis=1)

# Display Dataframe and shape
display(X_test_encoded.head())
display(X_test_encoded.shape)

# Check to see if dataframes contain any nulls
#display(X_test_encoded.isnull().sum())

Age  Education  JobSatisfaction  PerformanceRating  TotalWorkingYears  \
0   28          3                1                  3                  6   
1   53          2                1                  3                  5   
2   24          1                3                  3                  1   
3   45          3                1                  3                 25   
4   36          2                2                  3                 16   

   TrainingTimesLastYear  YearsAtCompany  YearsInCurrentRole  \
0                      4               5                   4   
1                      3               4                   2   
2                      2               1                   0   
3                      2               1                   0   
4                      3              13                  11   

   YearsSinceLastPromotion  EducationField_Life Sciences  ...  \
0                        1                           0.0  ...   
1                        1                           0.0  ...   
2                        0                           0.0  ...   
3                        0                           1.0  ...   
4                        3                           1.0  ...   

   EducationField_Other  EducationField_Technical Degree  \
0                   0.0                              0.0   
1                   0.0                              0.0   
2                   0.0                              0.0   
3                   0.0                              0.0   
4                   0.0                              0.0   

   JobRole_Human Resources  JobRole_Laboratory Technician  JobRole_Manager  \
0                      0.0                            0.0              0.0   
1                      0.0                            0.0              0.0   
2                      1.0                            0.0              0.0   
3                      0.0                            0.0              0.0   
4                      0.0                            1.0              0.0   

   JobRole_Manufacturing Director  JobRole_Research Director  \
0                             0.0                        0.0   
1                             1.0                        0.0   
2                             0.0                        0.0   
3                             0.0                        0.0   
4                             0.0                        0.0   

   JobRole_Research Scientist  JobRole_Sales Executive  \
0                         0.0                      1.0   
1                         0.0                      0.0   
2                         0.0                      0.0   
3                         1.0                      0.0   
4                         0.0                      0.0   

   JobRole_Sales Representative  
0                           0.0  
1                           0.0  
2                           0.0  
3                           0.0  
4                           0.0  

[5 rows x 22 columns]

(368, 22)

In [11]:
# Display X_train and X_test datatypes
display(X_train_encoded.dtypes)
display(X_test_encoded.dtypes)

Age                                  int64
Education                            int64
JobSatisfaction                      int64
PerformanceRating                    int64
TotalWorkingYears                    int64
TrainingTimesLastYear                int64
YearsAtCompany                       int64
YearsInCurrentRole                   int64
YearsSinceLastPromotion              int64
EducationField_Life Sciences       float64
EducationField_Marketing           float64
EducationField_Medical             float64
EducationField_Other               float64
EducationField_Technical Degree    float64
JobRole_Human Resources            float64
JobRole_Laboratory Technician      float64
JobRole_Manager                    float64
JobRole_Manufacturing Director     float64
JobRole_Research Director          float64
JobRole_Research Scientist         float64
JobRole_Sales Executive            float64
JobRole_Sales Representative       float64
dtype: object

Age                                  int64
Education                            int64
JobSatisfaction                      int64
PerformanceRating                    int64
TotalWorkingYears                    int64
TrainingTimesLastYear                int64
YearsAtCompany                       int64
YearsInCurrentRole                   int64
YearsSinceLastPromotion              int64
EducationField_Life Sciences       float64
EducationField_Marketing           float64
EducationField_Medical             float64
EducationField_Other               float64
EducationField_Technical Degree    float64
JobRole_Human Resources            float64
JobRole_Laboratory Technician      float64
JobRole_Manager                    float64
JobRole_Manufacturing Director     float64
JobRole_Research Director          float64
JobRole_Research Scientist         float64
JobRole_Sales Executive            float64
JobRole_Sales Representative       float64
dtype: object

In [12]:
# Display X_train and X_test columns
display(X_train_encoded.columns)
display(X_test_encoded.columns)

Index(['Age', 'Education', 'JobSatisfaction', 'PerformanceRating',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany',
       'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'EducationField_Life Sciences', 'EducationField_Marketing',
       'EducationField_Medical', 'EducationField_Other',
       'EducationField_Technical Degree', 'JobRole_Human Resources',
       'JobRole_Laboratory Technician', 'JobRole_Manager',
       'JobRole_Manufacturing Director', 'JobRole_Research Director',
       'JobRole_Research Scientist', 'JobRole_Sales Executive',
       'JobRole_Sales Representative'],
      dtype='object')

Index(['Age', 'Education', 'JobSatisfaction', 'PerformanceRating',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany',
       'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'EducationField_Life Sciences', 'EducationField_Marketing',
       'EducationField_Medical', 'EducationField_Other',
       'EducationField_Technical Degree', 'JobRole_Human Resources',
       'JobRole_Laboratory Technician', 'JobRole_Manager',
       'JobRole_Manufacturing Director', 'JobRole_Research Director',
       'JobRole_Research Scientist', 'JobRole_Sales Executive',
       'JobRole_Sales Representative'],
      dtype='object')

In [13]:
# Create a StandardScaler
std_scaler = StandardScaler()
cols_to_scale = ['Age', 'Education', 'JobSatisfaction',
                 'TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany',
                 'YearsInCurrentRole', 'YearsSinceLastPromotion']

# Fit and transform the StandardScaler to the training data
X_train_encoded_scaled = X_train_encoded[cols_to_scale]
X_train_scaled = std_scaler.fit_transform(X_train_encoded_scaled)
#display(X_train_scaled)
df_X_train_scaled = pd.DataFrame(X_train_scaled, columns=cols_to_scale)
display(df_X_train_scaled)

# Scale the testing data
X_test_encoded_scaled = X_test_encoded[cols_to_scale]
X_test_scaled = std_scaler.transform(X_test_encoded_scaled)
#display(X_test_scaled)
df_X_test_scaled = pd.DataFrame(X_test_scaled, columns=cols_to_scale)
display(df_X_test_scaled)


Age  Education  JobSatisfaction  TotalWorkingYears  \
0    -0.852159   0.094018        -1.569331          -0.014296   
1    -0.093088   1.053349         0.251093           0.498037   
2    -0.309965   0.094018        -1.569331           0.498037   
3    -1.069036   0.094018         1.161305          -0.526629   
4    -0.526843   0.094018         1.161305          -0.142379   
...        ...        ...              ...                ...   
1097 -0.201527   0.094018         0.251093          -0.142379   
1098  0.449105   0.094018        -0.659119          -0.014296   
1099 -1.611229   1.053349         1.161305          -1.295129   
1100 -0.852159  -0.865313        -0.659119          -0.142379   
1101  1.425053   0.094018         0.251093           2.035037   

      TrainingTimesLastYear  YearsAtCompany  YearsInCurrentRole  \
0                 -0.615111       -0.642460           -0.603137   
1                  1.695749       -0.970918           -1.155867   
2                 -0.615111        0.999830            1.331416   
3                  1.695749        0.014456            0.778687   
4                 -0.615111        0.507143            0.778687   
...                     ...             ...                 ...   
1097               0.155176        0.507143            1.331416   
1098               0.155176       -0.642460           -0.603137   
1099               1.695749       -1.135147           -1.155867   
1100              -0.615111       -0.478231           -0.326772   
1101              -0.615111       -0.970918           -1.155867   

      YearsSinceLastPromotion  
0                   -0.360548  
1                   -0.682269  
2                    0.282892  
3                   -0.682269  
4                   -0.682269  
...                       ...  
1097                 1.248052  
1098                -0.360548  
1099                -0.682269  
1100                -0.682269  
1101                -0.682269  

[1102 rows x 8 columns]

Age  Education  JobSatisfaction  TotalWorkingYears  \
0   -0.960597   0.094018        -1.569331          -0.654713   
1    1.750369  -0.865313        -1.569331          -0.782796   
2   -1.394352  -1.824643         0.251093          -1.295129   
3    0.882860   0.094018        -1.569331           1.778870   
4   -0.093088  -0.865313        -0.659119           0.626121   
..        ...        ...              ...                ...   
363 -1.177475  -0.865313         1.161305          -0.782796   
364  0.015351   0.094018        -0.659119          -0.398546   
365  0.340667  -0.865313         1.161305          -0.270463   
366 -1.285913   0.094018        -0.659119          -0.654713   
367  1.099737   1.053349        -0.659119           0.626121   

     TrainingTimesLastYear  YearsAtCompany  YearsInCurrentRole  \
0                 0.925462       -0.314002           -0.050408   
1                 0.155176       -0.478231           -0.603137   
2                -0.615111       -0.970918           -1.155867   
3                -0.615111       -0.970918           -1.155867   
4                 0.155176        0.999830            1.884145   
..                     ...             ...                 ...   
363              -0.615111       -0.806689           -0.603137   
364              -0.615111        0.178685           -0.050408   
365               0.155176        0.342914            1.055051   
366               0.155176       -0.806689           -1.155867   
367               0.925462        0.178685            0.778687   

     YearsSinceLastPromotion  
0                  -0.360548  
1                  -0.360548  
2                  -0.682269  
3                  -0.682269  
4                   0.282892  
..                       ...  
363                -0.038828  
364                 1.569772  
365                 0.604612  
366                -0.360548  
367                -0.360548  

[368 rows x 8 columns]

In [14]:
# Create a LabelEncoder
# Label encoding for binary; one-hot encoding for multiple categories

pr_encoder = LabelEncoder()
# Column to Label Encode 'PerformanceRating'

# Fit_transform the LabelEncoder to the training data
df_X_train_scaled['PerformanceRating'] = pr_encoder.fit_transform(X_train_encoded[['PerformanceRating']])
display(df_X_train_scaled)
display(df_X_train_scaled.shape)
#display(df_X_train_scaled['PerformanceRating'].value_counts())

# LabelEncode to the test data
df_X_test_scaled['PerformanceRating'] = pr_encoder.transform(X_test_encoded[['PerformanceRating']])
display(df_X_test_scaled)
display(df_X_test_scaled.shape)
#display(df_X_test_scaled['PerformanceRating'].value_counts())


/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:116: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Age  Education  JobSatisfaction  TotalWorkingYears  \
0    -0.852159   0.094018        -1.569331          -0.014296   
1    -0.093088   1.053349         0.251093           0.498037   
2    -0.309965   0.094018        -1.569331           0.498037   
3    -1.069036   0.094018         1.161305          -0.526629   
4    -0.526843   0.094018         1.161305          -0.142379   
...        ...        ...              ...                ...   
1097 -0.201527   0.094018         0.251093          -0.142379   
1098  0.449105   0.094018        -0.659119          -0.014296   
1099 -1.611229   1.053349         1.161305          -1.295129   
1100 -0.852159  -0.865313        -0.659119          -0.142379   
1101  1.425053   0.094018         0.251093           2.035037   

      TrainingTimesLastYear  YearsAtCompany  YearsInCurrentRole  \
0                 -0.615111       -0.642460           -0.603137   
1                  1.695749       -0.970918           -1.155867   
2                 -0.615111        0.999830            1.331416   
3                  1.695749        0.014456            0.778687   
4                 -0.615111        0.507143            0.778687   
...                     ...             ...                 ...   
1097               0.155176        0.507143            1.331416   
1098               0.155176       -0.642460           -0.603137   
1099               1.695749       -1.135147           -1.155867   
1100              -0.615111       -0.478231           -0.326772   
1101              -0.615111       -0.970918           -1.155867   

      YearsSinceLastPromotion  PerformanceRating  
0                   -0.360548                  0  
1                   -0.682269                  0  
2                    0.282892                  0  
3                   -0.682269                  0  
4                   -0.682269                  0  
...                       ...                ...  
1097                 1.248052                  0  
1098                -0.360548                  0  
1099                -0.682269                  0  
1100                -0.682269                  0  
1101                -0.682269                  0  

[1102 rows x 9 columns]

(1102, 9)

/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_label.py:134: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, dtype=self.classes_.dtype, warn=True)


Age  Education  JobSatisfaction  TotalWorkingYears  \
0   -0.960597   0.094018        -1.569331          -0.654713   
1    1.750369  -0.865313        -1.569331          -0.782796   
2   -1.394352  -1.824643         0.251093          -1.295129   
3    0.882860   0.094018        -1.569331           1.778870   
4   -0.093088  -0.865313        -0.659119           0.626121   
..        ...        ...              ...                ...   
363 -1.177475  -0.865313         1.161305          -0.782796   
364  0.015351   0.094018        -0.659119          -0.398546   
365  0.340667  -0.865313         1.161305          -0.270463   
366 -1.285913   0.094018        -0.659119          -0.654713   
367  1.099737   1.053349        -0.659119           0.626121   

     TrainingTimesLastYear  YearsAtCompany  YearsInCurrentRole  \
0                 0.925462       -0.314002           -0.050408   
1                 0.155176       -0.478231           -0.603137   
2                -0.615111       -0.970918           -1.155867   
3                -0.615111       -0.970918           -1.155867   
4                 0.155176        0.999830            1.884145   
..                     ...             ...                 ...   
363              -0.615111       -0.806689           -0.603137   
364              -0.615111        0.178685           -0.050408   
365               0.155176        0.342914            1.055051   
366               0.155176       -0.806689           -1.155867   
367               0.925462        0.178685            0.778687   

     YearsSinceLastPromotion  PerformanceRating  
0                  -0.360548                  0  
1                  -0.360548                  0  
2                  -0.682269                  0  
3                  -0.682269                  0  
4                   0.282892                  0  
..                       ...                ...  
363                -0.038828                  0  
364                 1.569772                  0  
365                 0.604612                  0  
366                -0.360548                  0  
367                -0.360548                  0  

[368 rows x 9 columns]

(368, 9)

In [15]:
# Combine all columns back together to create X_train and X_test

# Create list of the OneHotEncoder columns
ohe_df_cols = ['EducationField_Life Sciences', 'EducationField_Marketing',
               'EducationField_Medical', 'EducationField_Other',
               'EducationField_Technical Degree', 'JobRole_Human Resources',
               'JobRole_Laboratory Technician', 'JobRole_Manager',
               'JobRole_Manufacturing Director', 'JobRole_Research Director',
               'JobRole_Research Scientist', 'JobRole_Sales Executive',
               'JobRole_Sales Representative']

# Create X_train and X_test dataframes with the OneHotEncoder columns
df_X_train_ohe = X_train_encoded[ohe_df_cols]
df_X_test_ohe = X_test_encoded[ohe_df_cols]

# Concatenate the encoded columns to the scaled X_train and X_test DataFrames
X_train_clean = pd.concat([df_X_train_scaled, df_X_train_ohe], axis=1)
display(X_train_clean)
display(X_train_clean.shape)

X_test_clean = pd.concat([df_X_test_scaled, df_X_test_ohe], axis=1)
display(X_test_clean)
display(X_train_clean.shape)


Age  Education  JobSatisfaction  TotalWorkingYears  \
0    -0.852159   0.094018        -1.569331          -0.014296   
1    -0.093088   1.053349         0.251093           0.498037   
2    -0.309965   0.094018        -1.569331           0.498037   
3    -1.069036   0.094018         1.161305          -0.526629   
4    -0.526843   0.094018         1.161305          -0.142379   
...        ...        ...              ...                ...   
1097 -0.201527   0.094018         0.251093          -0.142379   
1098  0.449105   0.094018        -0.659119          -0.014296   
1099 -1.611229   1.053349         1.161305          -1.295129   
1100 -0.852159  -0.865313        -0.659119          -0.142379   
1101  1.425053   0.094018         0.251093           2.035037   

      TrainingTimesLastYear  YearsAtCompany  YearsInCurrentRole  \
0                 -0.615111       -0.642460           -0.603137   
1                  1.695749       -0.970918           -1.155867   
2                 -0.615111        0.999830            1.331416   
3                  1.695749        0.014456            0.778687   
4                 -0.615111        0.507143            0.778687   
...                     ...             ...                 ...   
1097               0.155176        0.507143            1.331416   
1098               0.155176       -0.642460           -0.603137   
1099               1.695749       -1.135147           -1.155867   
1100              -0.615111       -0.478231           -0.326772   
1101              -0.615111       -0.970918           -1.155867   

      YearsSinceLastPromotion  PerformanceRating  \
0                   -0.360548                  0   
1                   -0.682269                  0   
2                    0.282892                  0   
3                   -0.682269                  0   
4                   -0.682269                  0   
...                       ...                ...   
1097                 1.248052                  0   
1098                -0.360548                  0   
1099                -0.682269                  0   
1100                -0.682269                  0   
1101                -0.682269                  0   

      EducationField_Life Sciences  ...  EducationField_Other  \
0                              1.0  ...                   0.0   
1                              1.0  ...                   0.0   
2                              0.0  ...                   1.0   
3                              0.0  ...                   0.0   
4                              0.0  ...                   0.0   
...                            ...  ...                   ...   
1097                           1.0  ...                   0.0   
1098                           1.0  ...                   0.0   
1099                           1.0  ...                   0.0   
1100                           0.0  ...                   1.0   
1101                           0.0  ...                   0.0   

      EducationField_Technical Degree  JobRole_Human Resources  \
0                                 0.0                      0.0   
1                                 0.0                      0.0   
2                                 0.0                      0.0   
3                                 0.0                      0.0   
4                                 0.0                      0.0   
...                               ...                      ...   
1097                              0.0                      0.0   
1098                              0.0                      0.0   
1099                              0.0                      0.0   
1100                              0.0                      0.0   
1101                              0.0                      0.0   

      JobRole_Laboratory Technician  JobRole_Manager  \
0                               1.0              0.0   
1                               0.0              0.0   
2                               0.0              0.0   
3 

(1102, 22)

Age  Education  JobSatisfaction  TotalWorkingYears  \
0   -0.960597   0.094018        -1.569331          -0.654713   
1    1.750369  -0.865313        -1.569331          -0.782796   
2   -1.394352  -1.824643         0.251093          -1.295129   
3    0.882860   0.094018        -1.569331           1.778870   
4   -0.093088  -0.865313        -0.659119           0.626121   
..        ...        ...              ...                ...   
363 -1.177475  -0.865313         1.161305          -0.782796   
364  0.015351   0.094018        -0.659119          -0.398546   
365  0.340667  -0.865313         1.161305          -0.270463   
366 -1.285913   0.094018        -0.659119          -0.654713   
367  1.099737   1.053349        -0.659119           0.626121   

     TrainingTimesLastYear  YearsAtCompany  YearsInCurrentRole  \
0                 0.925462       -0.314002           -0.050408   
1                 0.155176       -0.478231           -0.603137   
2                -0.615111       -0.970918           -1.155867   
3                -0.615111       -0.970918           -1.155867   
4                 0.155176        0.999830            1.884145   
..                     ...             ...                 ...   
363              -0.615111       -0.806689           -0.603137   
364              -0.615111        0.178685           -0.050408   
365               0.155176        0.342914            1.055051   
366               0.155176       -0.806689           -1.155867   
367               0.925462        0.178685            0.778687   

     YearsSinceLastPromotion  PerformanceRating  EducationField_Life Sciences  \
0                  -0.360548                  0                           0.0   
1                  -0.360548                  0                           0.0   
2                  -0.682269                  0                           0.0   
3                  -0.682269                  0                           1.0   
4                   0.282892                  0                           1.0   
..                       ...                ...                           ...   
363                -0.038828                  0                           1.0   
364                 1.569772                  0                           0.0   
365                 0.604612                  0                           0.0   
366                -0.360548                  0                           0.0   
367                -0.360548                  0                           0.0   

     ...  EducationField_Other  EducationField_Technical Degree  \
0    ...                   0.0                              0.0   
1    ...                   0.0                              0.0   
2    ...                   0.0                              0.0   
3    ...                   0.0                              0.0   
4    ...                   0.0                              0.0   
..   ...                   ...                              ...   
363  ...                   0.0                              0.0   
364  ...                   0.0                              0.0   
365  ...                   0.0                              1.0   
366  ...                   0.0                              0.0   
367  ...                   0.0                              1.0   

     JobRole_Human Resources  JobRole_Laboratory Technician  JobRole_Manager  \
0                        0.0                            0.0              0.0   
1                        0.0                            0.0              0.0   
2                        1.0                            0.0              0.0   
3                        0.0                            0.0              0.0   
4                        0.0                            1.0              0.0   
..                       ...                            ...              ...   
363                      0.0                            1.0              0.0   
364                      0.0  

(1102, 22)

In [16]:
# Display y_train column values
for cols in y_train.columns:
  print(f"-------------- y_train {cols} --------------")
  print(y_train[cols].unique())
  print("--------------------------------------------")
  print("\n")
# Display y_test column values
for cols in y_test.columns:
  print(f"-------------- y_test {cols} --------------")
  print(y_test[cols].unique())
  print("--------------------------------------------")
  print("\n")

-------------- y_train Attrition --------------
['No' 'Yes']
--------------------------------------------


-------------- y_train Department --------------
['Research & Development' 'Sales' 'Human Resources']
--------------------------------------------


-------------- y_test Attrition --------------
['No' 'Yes']
--------------------------------------------


-------------- y_test Department --------------
['Sales' 'Research & Development' 'Human Resources']
--------------------------------------------




In [17]:
# Create a OneHotEncoder for the Department column

# Get value counts for Department values:  'Sales', 'Research & Development', 'Human Resources'
display(y_train['Department'].value_counts())

# Create a OneHotEncoder for the Department column
department_encoder = OneHotEncoder(drop='first', sparse_output=False)

# Fit_transform the encoder to the training data
#y_train_reindex = y_train.reset_index(drop=True)
y_train_department_encoded = department_encoder.fit_transform(y_train[['Department']])
department_columns = department_encoder.get_feature_names_out(['Department'])

# Create new variable by applying the encoder
# to the training data
y_train_department = pd.DataFrame(y_train_department_encoded, columns=department_columns)
display(y_train_department.head())

# Transform the encoder to the testing data
#y_test_reindex = y_test.reset_index(drop=True)
y_test_department_encoded = department_encoder.transform(y_test[['Department']])

# Create new variable by applying the encoder
# to the testing data
y_test_department = pd.DataFrame(y_test_department_encoded, columns=department_columns)
display(y_test_department.head())


Research & Development    721
Sales                     336
Human Resources            45
Name: Department, dtype: int64

Department_Research & Development  Department_Sales
0                                1.0               0.0
1                                0.0               1.0
2                                0.0               1.0
3                                0.0               1.0
4                                0.0               1.0

Department_Research & Development  Department_Sales
0                                0.0               1.0
1                                1.0               0.0
2                                0.0               0.0
3                                1.0               0.0
4                                1.0               0.0

In [18]:
# Create a OneHotEncoder for the Attrition column

# Get value counts for Attrition values:  'No' 'Yes'
display(y_train['Attrition'].value_counts())

# Create a OneHotEncoder for the Attrition column
attrition_encoder = OneHotEncoder(drop='if_binary', sparse_output=False)

# Fit_transform the encoder to the training data
y_train_attrition_encoded = attrition_encoder.fit_transform(y_train[['Attrition']])
attrition_columns = attrition_encoder.get_feature_names_out(['Attrition'])

# Create new variable by applying the encoder
# to the training data
y_train_attrition = pd.DataFrame(y_train_attrition_encoded, columns=attrition_columns)
display(y_train_attrition.head())

# Transform the encoder to the testing data
y_test_attrition_encoded = attrition_encoder.transform(y_test[['Attrition']])

# Create new variable by applying the encoder
# to the testing data
y_test_attrition = pd.DataFrame(y_test_attrition_encoded, columns=attrition_columns)
display(y_test_attrition.head())


No     913
Yes    189
Name: Attrition, dtype: int64

Attrition_Yes
0            0.0
1            0.0
2            0.0
3            0.0
4            0.0

Attrition_Yes
0            0.0
1            0.0
2            1.0
3            0.0
4            0.0

## Create, Compile, and Train the Model

In [19]:
# Find the number of columns in the X training data
X_cols = X_train_clean.shape[1]
#display(X_cols)

# Create the input layer
input_layer = layers.Input(shape=(X_cols,), name='input_features')

# Create at least two shared layers
shared_layer1 = layers.Dense(64, activation='relu')(input_layer)
shared_layer2 = layers.Dense(128, activation='relu')(shared_layer1)

In [20]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
department_dense = layers.Dense(32, activation='relu')(shared_layer2)

# Create the output layer
department_output = layers.Dense(len(y_train_department.columns),
                                 activation='softmax',
                                 name='department_output')(department_dense)


In [21]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attrition_dense = layers.Dense(32, activation='relu')(shared_layer2)

# Create the output layer
attrition_output = layers.Dense(len(y_train_attrition.columns),
                             activation='sigmoid',
                             name='attrition_output')(attrition_dense)


In [22]:
# Create the model
model = Model(inputs=input_layer, outputs=[department_output, attrition_output])

# Compile the model
model.compile(optimizer='adam',
              loss={'department_output': 'categorical_crossentropy', 'attrition_output': 'binary_crossentropy'},
              metrics={'department_output': 'accuracy', 'attrition_output': 'accuracy'})


# Summarize the model
model.summary()

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_features (InputLayer  [(None, 22)]                 0         []                            
 )                                                                                                
                                                                                                  
 dense (Dense)               (None, 64)                   1472      ['input_features[0][0]']      
                                                                                                  
 dense_1 (Dense)             (None, 128)                  8320      ['dense[0][0]']               
                                                                                                  
 dense_2 (Dense)             (None, 32)                   4128      ['dense_1[0][0]']         

In [23]:
# Train the model
model.fit(
    X_train_clean,
    {
        'department_output': y_train_department,
        'attrition_output': y_train_attrition
    },
    epochs=100,
    batch_size=32,
    validation_split=0.2
)


Epoch 1/100
28/28 [==============================] - 5s 50ms/step - loss: 1.1050 - department_output_loss: 0.5579 - attrition_output_loss: 0.5471 - department_output_accuracy: 0.6992 - attrition_output_accuracy: 0.7526 - val_loss: 1.0099 - val_department_output_loss: 0.4973 - val_attrition_output_loss: 0.5127 - val_department_output_accuracy: 0.6787 - val_attrition_output_accuracy: 0.7873
Epoch 2/100
28/28 [==============================] - 0s 8ms/step - loss: 0.8000 - department_output_loss: 0.3558 - attrition_output_loss: 0.4442 - department_output_accuracy: 0.8502 - attrition_output_accuracy: 0.8388 - val_loss: 0.7223 - val_department_output_loss: 0.2348 - val_attrition_output_loss: 0.4875 - val_department_output_accuracy: 0.9231 - val_attrition_output_accuracy: 0.7873
Epoch 3/100
28/28 [==============================] - 0s 14ms/step - loss: 0.5618 - department_output_loss: 0.1455 - attrition_output_loss: 0.4163 - department_output_accuracy: 0.9432 - attrition_output_accuracy: 0.838

In [24]:
# Evaluate the model with the testing data
test_results = model.evaluate(X_test_clean, {'department_output': y_test_department, 'attrition_output': y_test_attrition})
display(model.metrics_names)
display(test_results)

12/12 [==============================] - 0s 3ms/step - loss: 72.4911 - department_output_loss: 30.9848 - attrition_output_loss: 41.5064 - department_output_accuracy: 0.7989 - attrition_output_accuracy: 0.8696


['loss',
 'department_output_loss',
 'attrition_output_loss',
 'department_output_accuracy',
 'attrition_output_accuracy']

[72.49114227294922,
 30.98477554321289,
 41.5063591003418,
 0.7989130616188049,
 0.8695651888847351]

In [25]:
# Print the accuracy for both department and attrition
print(f"Department Accuracy: {test_results[3]}")
print(f"Attrition Accuracy: {test_results[4]}")

Department Accuracy: 0.7989130616188049
Attrition Accuracy: 0.8695651888847351


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. Is accuracy the best metric to use on this data? Why or why not?
  - Accuracy is appropriate for measuring Department predictions because the data has some balance, R&D (65%), Sales (30%), and HR (4%).  The accuracy score of 80% seems manageable for predicting a department that may be a better fit for the employee.  Also, the cost of incorrectly predicting a new department is low because you can engage the employee in a career developemnt discussion to assess their interest in switching to a new department and establish the best approach for a potential career change.
  - Precision may be better for measuring Attrition predictions because the data is imbalanced, No (83%), Yes (17%).  This is a contributing factor to the high accuracy score of 87%.  Also, there may be a high cost to incorrectly predicting that an employee is going to stay, but they actually leave.  Companies make investments in their employees to build their knowledge and develop relevant skills.  They want to keep them if they can.
2.  What activation functions did you choose for your output layers, and why?   
  - **department_output:**  The model predicts 3 departments.  I used Softmax because it is recommended for mulit-class classification.  
  - **attribution_output:**  The model predicts a binary Yes/No attibution.  I used Sigmoid because it is recommended for binary classification.

3. Can you name a few ways that this model might be improved?
  - Add or Reduce the number of inputs
  - Increase the number of layers
  - Increase the number of neurons
  - Increase the number of epochs
  - Consider splitting this into separate models to better focus models on predicting employee attrition and departments better suited for them.